Sphinx documentation

wigging opened the pull request:
This adds Sphinx for generating documentation for the package. The Furo theme is also added for a more modern design and layout of the documentation pages. The associated dependencies are added to the optional `docs` group in the `pyproject.toml` file.

wigging commented:
I will rebase this with the dev branch after the other pull request has been merged. The number commits for this pull request will be much lower after the rebase.

wigging commented:
Looks like I need to update the CONTRIBUTING.md with info on how to build the docs. I'll do that soon.

wigging commented:
@renan-souza I added a few lines to the contributing document. I'll focus on resolving that issue in a separate pull request.

renan-souza commented:
> Looks like I need to update the CONTRIBUTING.md with info on how to build the docs. I'll do that soon.

Since you are changing it in this PR, do you mind adjusting the text that says we use black/flake for code formatting?

wigging commented:
Black and Flake8 are not used in this project.

renan-souza commented:
> Black and Flake8 are not used in this project.

Exactly. But the CONTRIBUTING says they are. Since you are updating the CONTRIBUTING, can you please update the text in it where it mentions those and update with what are currently using for code formating?

wigging commented:
Cleaned up the contributing guidelines and added more info about code linting and formatting.

wigging commented:
@renan-souza I don't have anything else to add to this pull request.

renan-souza commented:
> @renan-souza I don't have anything else to add to this pull request.

Sounds good. I will start reviewing it and add new commits if needed.

renan-souza commented on docs/index.rst:
As is this is only for improvement thoughts, there is no action needed right now in this PR for my following comment:

Is it possible to centralize all these doc-related texts into one place (e.g., a file) and fill in both the markdowns and rst from them? Or, get these texts in these rst from the markdown files?

I like to avoid duplicating things, especially text and code, as it is very easy to update one and forget the other.

wigging replied:
I just copied and pasted the README as temporary solution. At some point the homepage of the Sphinx docs will be different than the README for the GitHub repository. So don't want to blindly copy the README but it can be done if necessary. We can figure it out as things progress.

renan-souza commented on Makefile:
I like the Makefile idea a lot. Thanks for bringing this up.

Two suggestions:

1) The Makefile itself is not being covered in any test. In the GH actions that run tests, we could use the makefile commands. This would assure that the makefile commands will always be up to date and working.

2) Please add a `test_all` command (or anything like this) that does not ignore the llm tests and the notebook tests. Something like:

```
test_all:
      pytest
      pytest --nbmake "notebooks/" --nbmake-timeout=600 --ignore=notebooks/dask_from_CLI.ipynb     
```

And this could be the command to use in the GH actions as well.

wigging replied:
I don't see any GitHub Actions workflows that run all the tests with `pytest`. They all ignore some part of tests. Submit an issue about using the Makefile with the GitHub Actions and we can discuss this further in that issue before making more changes.

renan-souza replied:
Sorry, I wasn't clear. 
The `test_all` (we need a better term because even this wouldn't test all, as it ignores the dask from cli test) would be to run locally.

The `test` command would be used in the GH actions that  test. In this case, we would need to add the notebook tests to the `test` command as well.

Sure, we can create a new issue/PR for this.

renan-souza commented:
@wigging Before we merge this PR, I would feel more comfortable if we tested the sphinx docs procedure. 

Do you mind adding a new CI workflow based on https://github.com/ORNL/flowcept/blob/sphinx-docs/.github/workflows/run-checks.yml to test if the documents are being generated correctly?

This workflow can run only on PRs, just like the run-checks.yml.

Also, you could create a new command in [Flowcept's Makefile](https://github.com/ORNL/flowcept/blob/sphinx-docs/Makefile) to generate the documents. 

Once we put sphinx as part of the CI process and it passes the test, I'll merge & close this PR.

wigging commented:
It builds fine on my local computer but I'll add something to the GitHub Actions to check the build too. And Sphinx already provides a Makefile (see docs/Makefile) for building the documentation and doing other things.

renan-souza commented:
> It builds fine on my local computer but I'll add something to the GitHub Actions to check the build too.

Thanks. The goal is to increase the test coverage of all these features we add into the project.

> And Sphinx already provides a Makefile (see docs/Makefile) for building the documentation and doing other things.
Yea, I saw that, that's fine, but since we now have our own Makefile, it's more convenient to use that same Makefile for everything rather having multiple Makefiles and cd into a directory to run it.

wigging commented:
> Yea, I saw that, that's fine, but since we now have our own Makefile, it's more convenient to use that same Makefile for everything rather having multiple Makefiles and cd into a directory to run it.

You can pursue this in a separate pull request. This pull request is about setting up Sphinx, not fiddling around with Makefiles. The Makefile in the docs directory is generated by Sphinx and it uses the underlying sphinx CLI commands. Those commands expect to be run from the root level of the docs directory. I don't know what all these commands are but you will have to avoid naming conflicts if you do a recursive Makefile. You also would need to properly handle the path to the sphinx config file and other files if you try to use a top-level Makefile. Just cd into the docs directory and use the Sphinx Makefile as intended.

renan-souza commented:
> @renan-souza Did something change on the dev branch to cause the tests to fail? None of the changes in this pull request should affect the tests.

Nothing has changed that could cause that error. Not sure what happened. I'll need to debug.

wigging commented:
All of the errors and test failures seem to be caused by code that uses Dask.

renan-souza commented:
> All of the errors and test failures seem to be caused by code that uses Dask.

Yes and nothing about Dask has changed on our side. 
Since the CI always install the newest version of these dependencies, I suspect that something changed in the newest version of Dask that caused this error.

Please ignore these Dask errors and they have nothing to do with your chages.

I'll need to dig into this later.

wigging commented:
I went ahead and added the Makefile commands to build/clean the Sphinx docs from the top-level of the project. I removed the Makefile in the docs directory. I also added the Makefile commands in the run-checks workflow.

wigging commented:
@renan-souza Anything else needed here?
